lib/spack/spack/binary_distribution.py

```diff
@@ -133,6 +131,349 @@ def __init__(self, errors):
         super().__init__(self.message)
 
 
+class ExistsInBuildcache(NamedTuple):
+    signed: bool
+    signed_url: str
+    unsigned: bool
+    unsigned_url: str
+    tarball: bool
+    tarball_url: str
+
+
+class URLBuildcacheEntry:
+    """A class for managing URL-style buildcache entries
+
+    This class manages access to a versioned buildcache entry by providing
+    a means to download both the metadata (spec file) and compressed archive.
+    It also provides methods for accessing the paths/urls associcated with
+    buildcache entries.
+
+    Starting with buildcache layout version 3, it is not possible to know
+    the full path to a compressed archive without either building it locally,
+    or else fetching and reading the metadata first.  This class provides api
+    for fetching the metadata, as well as fetching the archive, and it enforces
+    the need to fetch the metadata first.
+
+    To help with downloading, this class manages two spack.spec.Stage objects
+    internally, which must be destroyed when finished.  Specifically, if you
+    call any of the following methods on an instance, you must eventually also
+    call destroy():
```

Comment:
Having to ensure `destroy()` is called makes me wonder if this would be better as a context object with a manager.

Reply:
I would prefer not to tackle making `URLBuildcacheEntry` a context manager in this PR, as I don't yet even see how it could be done. I agree it could be a nice improvement making it easier for developers to work with the class, I just don't think I'm ready to bite it off in this PR.